File: his/his/report/department_detail_profit_and_loss/department_detail_profit_and_loss.js
// Copyright (c) 2026, Rasiin Tech and contributors
// For license information, please see license.txt
/* eslint-disable */

const DETAIL_REPORT_VIEWS = {
	Department: {
		filter_field: "department",
	},
	Consultant: {
		filter_field: "consultant",
	},
};

function toggle_detail_filters() {
	const report = frappe.query_report;
	const view_by = report.get_filter_value("view_by") || "Department";
	const active_filter = DETAIL_REPORT_VIEWS[view_by].filter_field;

	["department", "consultant"].forEach((fieldname) => {
		const filter = report.get_filter(fieldname);
		if (!filter) return;

		const is_active = fieldname === active_filter;
		filter.df.hidden = !is_active;
		filter.df.reqd = is_active ? 1 : 0;

		if (!is_active) {
			filter.set_value("");
		}

		filter.refresh();
		filter.$wrapper.toggle(is_active);
	});
}

frappe.query_reports["Department Detail Profit and Loss"] = {
	filters: [
		{
			fieldname: "view_by",
			label: "View By",
			fieldtype: "Select",
			options: "Department\nConsultant",
			default: "Department",
			reqd: 1,
			on_change: function() {
				toggle_detail_filters();
				frappe.query_report.refresh();
			}
		},
		{
			fieldname: "department",
			label: "Department",
			fieldtype: "Link",
			options: "Medical Department",
			reqd: 1
		},
		{
			fieldname: "consultant",
			label: "Consultant",
			fieldtype: "Link",
			options: "Healthcare Practitioner",
			hidden: 1
		},
		{
			fieldname: "from_date",
			label: "From Date",
			fieldtype: "Date",
			reqd: 1
		},
		{
			fieldname: "to_date",
			label: "To Date",
			fieldtype: "Date",
			reqd: 1
		}
	],

	onload: function(report) {
		toggle_detail_filters();
	},

	formatter: function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (!data) return value;

		const view_by = frappe.query_report.get_filter_value("view_by") || "Department";
		const header_value = frappe.query_report.get_filter_value(
			DETAIL_REPORT_VIEWS[view_by].filter_field
		);

		if (
			data.entity === header_value ||
			["Income", "Expenses", "Direct Expense", "Indirect Expense", "Total Expense", "Net Profit & Loss"].includes(data.entity)
		) {
			return `<b>${value || ""}</b>`;
		}

		if (
			column.fieldname === "amount" &&
			parseFloat(data.amount || 0) < 0
		) {
			return `<span style="color:red;font-weight:600">${value || ""}</span>`;
		}

		if (
			column.fieldname === "amount" &&
			data.entity === "Net Profit & Loss" &&
			parseFloat(data.amount || 0) > 0
		) {
			return `<span style="color:green;font-weight:700">${value || ""}</span>`;
		}

		return value;
	}
};

File: his/his/report/department_detail_profit_and_loss/department_detail_profit_and_loss.py
# Copyright (c) 2026, Rasiin Tech and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import flt


VIEW_CONFIG = {
    "Department": {
        "fieldname": "department",
        "column_label": "Department",
        "filter_label": "Department",
    },
    "Consultant": {
        "fieldname": "consultant",
        "column_label": "Consultant",
        "filter_label": "Consultant",
    },
}


def execute(filters=None):
    filters = filters or {}
    view_by = filters.get("view_by") or "Department"

    if view_by not in VIEW_CONFIG:
        frappe.throw("Please select a valid View By option")

    filter_field = VIEW_CONFIG[view_by]["fieldname"]
    filter_value = filters.get(filter_field)

    if not filter_value:
        frappe.throw(f"Please select a {VIEW_CONFIG[view_by]['filter_label']}")

    columns = get_columns(view_by)
    data = get_data(filters, view_by, filter_value)
    return columns, data


def get_columns(view_by):
    return [
        {
            "label": VIEW_CONFIG[view_by]["column_label"],
            "fieldname": "entity",
            "fieldtype": "Data",
            "width": 220,
        },
        {
            "label": "Account",
            "fieldname": "account",
            "fieldtype": "Data",
            "width": 220,
        },
        {
            "label": "Amount",
            "fieldname": "amount",
            "fieldtype": "Currency",
            "width": 140,
        },
    ]


def get_data(filters, view_by, filter_value):
    filter_field = VIEW_CONFIG[view_by]["fieldname"]
    conditions = [f"c.{filter_field} = %(filter_value)s"]
    values = {"filter_value": filter_value}

    if filters.get("from_date"):
        conditions.append("p.from_date >= %(from_date)s")
        values["from_date"] = filters.get("from_date")

    if filters.get("to_date"):
        conditions.append("p.to_date <= %(to_date)s")
        values["to_date"] = filters.get("to_date")

    rows = frappe.db.sql(
        f"""
        SELECT
            p.type,
            p.account,
            SUM(c.allocatated_amount) AS amount
        FROM `tabDepartment Wise Profit and Loss` p
        INNER JOIN `tabAllocation Table` c
            ON c.parent = p.name
        WHERE p.docstatus < 2
          AND {' AND '.join(conditions)}
        GROUP BY p.type, p.account
        ORDER BY
            CASE
                WHEN p.type = 'Income' THEN 1
                WHEN p.type = 'Direct Expense' THEN 2
                WHEN p.type = 'Indirect Expense' THEN 3
                ELSE 4
            END,
            p.account
    """,
        values,
        as_dict=True,
    )

    data = []
    direct_income_total = 0
    direct_expense_total = 0
    indirect_expense_total = 0

    income_rows = []
    direct_expense_rows = []
    indirect_expense_rows = []

    for row in rows:
        amount = flt(row.amount)

        if row.type == "Income":
            income_rows.append({"entity": "", "account": row.account, "amount": amount})
            direct_income_total += amount
        elif row.type == "Direct Expense":
            direct_expense_rows.append({"entity": "", "account": row.account, "amount": amount})
            direct_expense_total += amount
        elif row.type == "Indirect Expense":
            indirect_expense_rows.append({"entity": "", "account": row.account, "amount": amount})
            indirect_expense_total += amount

    total_expense = direct_expense_total + indirect_expense_total
    net_profit = direct_income_total - total_expense

    data.append({"entity": filter_value, "account": "", "amount": ""})
    data.append({"entity": "Income", "account": "", "amount": ""})
    data.extend(income_rows)

    data.append({"entity": "Expenses", "account": "", "amount": ""})
    data.append({"entity": "Direct Expense", "account": "", "amount": ""})
    data.extend(direct_expense_rows)

    data.append({"entity": "Indirect Expense", "account": "", "amount": ""})
    data.extend(indirect_expense_rows)

    data.append({"entity": "Total Expense", "account": "", "amount": total_expense})
    data.append({})
    data.append({"entity": "Net Profit & Loss", "account": "", "amount": net_profit})

    return data
